ledger/participant-integration-api/src/main/scala/platform/store/backend/m/MStore.scala

```diff
@@ -0,0 +1,98 @@
+// TODO purge memory in the before after for persistence testing
+// TODO perfopt non compliant: drop the string interning population?
+// TODO perfopt non compliant: drop pruning?
+
+case class MStore(
+    ledgerEnd: ParameterStorageBackend.LedgerEnd = null,
+    identityParams: ParameterStorageBackend.IdentityParams = null,
+    prunedUpToInclusive: Offset = null,
+    allDivulgedContractsPrunedUpToInclusive: Offset = null,
+    events: Vector[DbDto] = Vector.empty,
+    contractIdIndex: Map[String, Vector[DbDto]] = Map.empty, // to all
+    contractKeyIndex: Map[String, Vector[DbDto]] = Map.empty, // to creates
+    transactionIdIndex: Map[String, Vector[DbDto]] = Map.empty, // to all
+
+    configurations: Vector[DbDto.ConfigurationEntry] = Vector.empty,
+    partyEntries: Vector[DbDto.PartyEntry] = Vector.empty,
+    partyIndex: Map[String, Vector[DbDto.PartyEntry]] = Map.empty,
+    packages: Map[String, DbDto.Package] = Map.empty,
+    packageEntries: Vector[DbDto.PackageEntry] = Vector.empty,
+    stringInternings: Vector[DbDto.StringInterningDto] = Vector.empty,
+    commandCompletions: Vector[DbDto.CommandCompletion] = Vector.empty,
+    meteringLedgerEnd: LedgerMeteringEnd = null,
+    transactionMetering: Vector[DbDto.TransactionMetering] = Vector.empty,
+    participantMetering: Vector[MeteringStore.ParticipantMetering] = Vector.empty,
+) {
+  def archived(ledgerEndSequentialId: Long)(dto: DbDto): Boolean =
+    contractIdIndex
+      .get(dto match {
+        case dto: DbDto.EventCreate => dto.contract_id
+        case dto: DbDto.EventDivulgence => dto.contract_id
+        case dto: DbDto.EventExercise => dto.contract_id
+        case _ => throw new Exception
+      })
+      .forall(_.reverseIterator.exists {
+        case archive: DbDto.EventExercise if archive.consuming =>
+          archive.event_sequential_id <= ledgerEndSequentialId
+        case _ => false
+      })
+
+  def eventRange(startExclusive: Long, endInclusive: Long): Iterator[DbDto] = {
+    val fromInclusive = startExclusive
+    // + 1 so it is inclusive
+    // - 1 so it is matching the Vector index (the first event_sequential_id is 1)
+    val toExclusive = endInclusive
+    // + 1 so it is exclusive
+    // - 1 so it is matching the Vector index (the first event_sequential_id is 1)
+
+    Range(fromInclusive.toInt, toExclusive.toInt).iterator
+      .filter(_ >= 0)
+      .filter(_ < events.size)
+      .map(events)
+  }
+}
+
+object MStore {
+  private class MStoreRef(@volatile var mStore: MStore)
+
+  private val globalMap: scala.collection.concurrent.Map[String, MStoreRef] =
+    new ConcurrentHashMap[String, MStoreRef]().asScala
+
+  def apply[T](connection: Connection)(f: MStore => T): T =
+    f(globalMap(idFrom(connection)).mStore)
+
+  def update(connection: Connection)(f: MStore => MStore): Unit = {
+    val ref = globalMap(idFrom(connection))
+    ref.synchronized {
+      ref.mStore = f(ref.mStore)
+    }
```

Comment:
`MStore` does not support any kind of transaction isolation: transaction-related operations on `MConnection` are all no-ops, and this `update` method (which corresponds to an individual statement within an SQL transaction) immediately modifies the data. Some thoughts about this:

- This _looks_ fine for the event store: all rows are immutable, and we use our own mechanism for preventing dirty reads (the ledger end pointers)
- How do we verify that this doesn't introduce some subtle concurrency bugs? For example, is it safe to read events while pruning is running? It took us a year to realize that H2 is broken.
- This corresponds to the `read uncommitted` isolation level. Would we feel comfortable executing all queries in `read uncommitted` isolation level (if we used a database that really supports this mode)?

Reply:
I would rather suggest to think about this as serializable for individual changes point of view (which it is, you cannot see any partial change which happens in a synchronized block, neither it is possible that these synchronized blocks interleave).
What we really should be on the lookout what happens between these serializable transactions. Pls raise all issues which you find discussion worthy!

Reply:
This is only "serializable" if each each SQL transaction (i.e., each `dbDispatcher.executeSql` call) contains at most one `StorageBackend` method call. Is that the case right now? If yes, how do we make sure it stays this way?

```scala
dbDispatcher.executeSql(conn => {
    storageBackend.ledgerEnd(conn)
    storageBackend.ledgerEnd(conn)
})
```
In this example, both calls will always return the same data with Postgres/Oracle, but may return different data with `MStore`.